### PR TITLE
Add validation

### PR DESCRIPTION
**Pull request by amitkano**
train_7.yaml can be used to run (small changes in default_config.py)

### CODE REVIEW

File: algorithms/dqn/trainer.py [hunk 1]
```diff
@@ -57,23 +61,42 @@ def train(self):
...
+                                                        q_validation_episodes_list,
+                                                        fig_num=3, y_label='Q value')
+
+            if (i_episode % self.cfg.VALIDATION.SCORE_VALIDATION_FREQUENCY == 0) and (i_episode > 0) and\
+                    (self.cfg.VALIDATION.SCORE_VALIDATION_FREQUENCY != -1):
+                validation_score = self.validate(val_type='score')
+                self.score_validation_score_list.append(validation_score)
+                score_validation_episodes_list.append(i_episode)
+                if self.cfg.VISUALIZE:
+
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Remove line

File: algorithms/dqn/trainer.py [hunk 2]
```diff
@@ -29,11 +29,14 @@ def __init__(self, train_cfg, env, agent, target_net, policy_net, memory, optimi
         self.policy_net = policy_net
         self.memory = memory
         self.steps_done = 0
+        self.init_episode = 0
         self.num_episodes = num_episodes
         self.device = device
         self.episode_durations = []
-        self.validation_score_list = []
-        self.env_state_list = env_state_list
+        self.q_validation_score_list = []
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: remove _list ending from names

File: algorithms/dqn/trainer.py [hunk 3]
```diff
@@ -172,13 +193,23 @@ def step(self):
...
 
-    def validate(self):
-        validation_scores = []
-        for state in self.env_state_list:
-            current_state_q = max(
+    def validate(self, val_type='q_value'):
+        validation_values = []
+        if val_type == 'q_value':
+            for state in self.env_random_state_list:
+                current_state_q = max(
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Add with torch.no_grad()

File: algorithms/dqn/trainer.py [hunk 4]
```diff
@@ -57,23 +61,42 @@ def train(self):
...
+            self._train_episode(current_screen)
             # Update the target network, copying all weights and biases in DQN
             if i_episode % self.cfg.TARGET_UPDATE == 0:
                 self.target_net.load_state_dict(self.agent.policy_net.state_dict())
 
-            if i_episode % self.cfg.VALIDATE_FREQUENCY == 0:
-                validation_score = self.validate()
-                self.validation_score_list.append(validation_score)
-                episodes_list.append(i_episode)
+            if (i_episode % self.cfg.VALIDATION.Q_VALIDATION_FREQUENCY == 0) and\
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Add mean values to log

File: algorithms/dqn/trainer.py [hunk 5]
```diff
@@ -227,12 +258,55 @@ def _get_newest_ckpt(self):
...
+    def __init__(self, policy_net, n_actions, device, env, epsilon=0.05):
         self.policy_net = policy_net
         self.state = None
         self.n_actions = n_actions
         self.device = device
+        self.env = env
+        self.epsilon = epsilon
+
+    def _play_episode(self, current_screen):
+        """
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Use google style docstring

File: algorithms/dqn/trainer.py [hunk 6]
```diff
@@ -227,12 +258,55 @@ def _get_newest_ckpt(self):
...
+            else:
+                next_state = None
+
+            # Store the transition in memory
+            # self.memory.push(self.agent.state, action, next_state, reward)
+
+            # Move to the next state
+            self.state = next_state
+
+            # Perform one step of the optimization (on the target network)
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Remove this comment

File: algorithms/dqn/trainer.py [hunk 7]
```diff
@@ -227,12 +258,55 @@ def _get_newest_ckpt(self):
             newest_ckpt_name = max(files, key=lambda x: int(x.split('.')[0].split('_')[-1]))
         return newest_ckpt_name
 
+
 class DQNAgent(object):
-    def __init__(self, policy_net, n_actions, device):
+    def __init__(self, policy_net, n_actions, device, env, epsilon=0.05):
```
Comment (maorshutman) [inline algorithms/dqn/trainer.py]: Set default eps to 0.0

File: algorithms/dqn/validation/validation_set_generator.py [hunk 8]
```diff
@@ -24,12 +24,34 @@
...
+EPISODES_TO_PLAY = 10
 SEED = 7
-PATH_TO_SAVE = '/raid/algo/SOCISP_SLOW/ADAS/Courses/mspacman/data/validation_set_random_' + str(VALIDATION_SET_SIZE) + '.pickle'
+PATH_TO_SAVE = '/raid/algo/SOCISP_SLOW/ADAS/Courses/mspacman/data/validation_set_initial_' + str(VALIDATION_SET_SIZE) + '.pickle'
 
 
 @ex.main
+def generate_init_states_list():
+    """
+    generates a validation random initial states and screens and saves it as .pickle
```
Comment (maorshutman) [inline algorithms/dqn/validation/validation_set_generator.py]: Capital letter and dot

File: algorithms/dqn/validation/validation_set_generator.py [hunk 9]
```diff
@@ -24,12 +24,34 @@
...
+    """
+    env = gym.make('CartPole-v0').unwrapped
+    env.seed(SEED)
+    env.reset()
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    init_state_list = []
+    init_screen_list = []
```
Comment (maorshutman) [inline algorithms/dqn/validation/validation_set_generator.py]: Add empty lines

File: utils/visualization.py [hunk 10]
```diff
@@ -22,13 +22,13 @@ def plot_durations(episode_durations):
...
+    plt.figure(fig_num)
     plt.clf()
     validation_score_list_t = torch.tensor(validation_score_list, dtype=torch.float)
     plt.title('Training...')
     plt.xlabel('Episode')
-    plt.ylabel('Q value')
+    plt.ylabel(y_label)
     plt.plot(episodes_list ,validation_score_list_t.numpy())
     # Take 100 episode averages and plot them too
     plt.pause(0.001)  # pause a bit so that plots are updated
```
Comment (maorshutman) [inline utils/visualization.py]: Check if neccesary

**Comment by amitkano** [comment]
fixed comments in pull request and pulled from master branch

**Pull request merged**